web/src/pages/index.tsx

```diff
@@ -70,6 +77,32 @@ const Home: NextPage = () => {
       }
   };
 
+
+  const updateCitations = (allCitations: Citation[], current: CurrentSearch) => {
```

Comment:
The LLM flow is:
1. various loading stages
2. citations are passed in
3. more loading phases
4. the main response starts coming in in small chunks
5. more loading
6. the followups are fetched
7. done

The main issue here is that the citations are loaded first, and there usually are like 10 possible articles to be cited, but they're not used until they appear in the actual text. And most of them won't be used anyway.
This means that magic has to be done to work out which of the possible citations are actually cited, and for them to then get the appropriate numbers.

I dislike this function. Quite a bit. But couldn't come up with a cleaner way of doing it :/
At least most of the magic is focused here, with subsequent code assuming that each citation will have an `index` field with the canonical number to be displayed